In [5]:
import numpy as np
import numpy.random as random
import scipy as sp
import pandas as pd
import sklearn
from pandas import Series,DataFrame

In [6]:
metadata=pd.read_csv("movies_metadata.csv")
metadata.head()

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview    ...     release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...       1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...       1995-12-15   
2  A family wedding reignites the ancient feud be...    ...       1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...       1995-12-22   
4  Just when George Banks has recovered from his ...    ...       1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [7]:
keywords=pd.read_csv("keywords.csv")
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

### データの抜き出し、整形

In [8]:
vote_average = metadata[["id","vote_average"]]
vote_average.head()

id  vote_average
0    862           7.7
1   8844           6.9
2  15602           6.5
3  31357           6.1
4  11862           5.7

In [9]:
#idがint型じゃない
vote_average[["id"]]=vote_average.id.convert_objects(convert_numeric=True).fillna(-1).astype(np.int)
vote_average.sort_values(by='id')

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  from ipykernel import kernelapp as app
/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.or

id  vote_average
19730      -1           NaN
29503      -1           NaN
35587      -1           NaN
4342        2           7.1
12947       3           7.1
17          5           6.5
474         6           6.4
256        11           8.1
6232       12           7.6
351        13           8.2
2742       14           7.9
896        15           8.0
3786       16           7.7
10765      17           5.6
1460       18           7.3
1901       19           8.0
6621       20           7.2
3531       21           7.8
6390       22           7.5
6725       24           7.7
10517      25           6.6
9769       26           6.5
10083      27           5.1
1165       28           8.0
38451      30           7.9
1221       33           7.7
11959      35           6.9
7208       38           7.9
4108       55           7.6
11008      58           7.0
...       ...           ...
44939  459928           5.1
45225  459950           7.3
44590  460024           4.0
45371  460135          10.0
44883  460822           6.8
45031  460846           6.0
21944  460870           0.0
45184  461053           5.2
44981  461088           0.0
44613  461089           0.0
45465  461257           0.0
45417  461297           6.0
44664  461533           0.0
44829  461615           7.5
45170  461634           8.0
44763  461805           8.1
44772  461955           7.4
44975  462026           0.0
45179  462108           5.8
23428  462788           0.0
44969  463800           7.3
45145  463906           5.8
45200  464111           7.2
45085  464207           7.5
45000  464819           0.0
45078  465044           0.0
45273  467731           0.0
21891  468343           0.0
45398  468707           8.0
20189  469172           0.0

[45466 rows x 2 columns]

In [10]:
vote_average.sort_values(by='id')

id  vote_average
19730      -1           NaN
29503      -1           NaN
35587      -1           NaN
4342        2           7.1
12947       3           7.1
17          5           6.5
474         6           6.4
256        11           8.1
6232       12           7.6
351        13           8.2
2742       14           7.9
896        15           8.0
3786       16           7.7
10765      17           5.6
1460       18           7.3
1901       19           8.0
6621       20           7.2
3531       21           7.8
6390       22           7.5
6725       24           7.7
10517      25           6.6
9769       26           6.5
10083      27           5.1
1165       28           8.0
38451      30           7.9
1221       33           7.7
11959      35           6.9
7208       38           7.9
4108       55           7.6
11008      58           7.0
...       ...           ...
44939  459928           5.1
45225  459950           7.3
44590  460024           4.0
45371  460135          10.0
44883  460822           6.8
45031  460846           6.0
21944  460870           0.0
45184  461053           5.2
44981  461088           0.0
44613  461089           0.0
45465  461257           0.0
45417  461297           6.0
44664  461533           0.0
44829  461615           7.5
45170  461634           8.0
44763  461805           8.1
44772  461955           7.4
44975  462026           0.0
45179  462108           5.8
23428  462788           0.0
44969  463800           7.3
45145  463906           5.8
45200  464111           7.2
45085  464207           7.5
45000  464819           0.0
45078  465044           0.0
45273  467731           0.0
21891  468343           0.0
45398  468707           8.0
20189  469172           0.0

[45466 rows x 2 columns]

### keywords_vote... id,keywords,vote_average

In [11]:
keywords_vote=pd.merge(keywords,vote_average,on="id")
keywords_vote.head()

id                                           keywords  vote_average
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7

### vote=0,NaN / keywords=[ ]のデータを除外

In [12]:
#NaN削除
keywords_vote.dropna()

id                                           keywords  vote_average
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7
5         949  [{'id': 642, 'name': 'robbery'}, {'id': 703, '...           7.7
6       11860  [{'id': 90, 'name': 'paris'}, {'id': 380, 'nam...           6.2
7       45325                                                 []           5.4
8        9091  [{'id': 949, 'name': 'terrorist'}, {'id': 1562...           5.5
9         710  [{'id': 701, 'name': 'cuba'}, {'id': 769, 'nam...           6.6
10       9087  [{'id': 833, 'name': 'white house'}, {'id': 84...           6.5
11      12110  [{'id': 3633, 'name': 'dracula'}, {'id': 11931...           5.7
12      21032  [{'id': 1994, 'name': 'wolf'}, {'id': 6411, 'n...           7.1
13      10858  [{'id': 840, 'name': 'usa president'}, {'id': ...           7.1
14       1408  [{'id': 911, 'name': 'exotic island'}, {'id': ...           5.7
15        524  [{'id': 383, 'name': 'poker'}, {'id': 726, 'na...           7.8
16       4584  [{'id': 420, 'name': 'bowling'}, {'id': 818, '...           7.2
17          5  [{'id': 612, 'name': 'hotel'}, {'id': 613, 'na...           6.5
18       9273  [{'id': 409, 'name': 'africa'}, {'id': 1551, '...           6.1
19      11517  [{'id': 380, 'name': 'brother brother relation...           5.4
20       8012  [{'id': 395, 'name': 'gambling'}, {'id': 416, ...           6.4
21       1710  [{'id': 796, 'name': 'police brutality'}, {'id...           6.5
22       9691  [{'id': 271, 'name': 'competition'}, {'id': 44...           6.0
23      12665  [{'id': 7345, 'name': 'dead animal'}, {'id': 1...           6.3
24        451  [{'id': 30, 'name': 'individual'}, {'id': 549,...           7.1
25      16420  [{'id': 497, 'name': 'shakespeare'}, {'id': 93...           7.0
26       9263  [{'id': 1228, 'name': '1970s'}, {'id': 5248, '...           6.6
27      17015  [{'id': 1316, 'name': 'captain'}, {'id': 33384...           7.4
28        902  [{'id': 402, 'name': 'clone'}, {'id': 1566, 'n...           7.6
29      37557  [{'id': 964, 'name': 'servant'}, {'id': 2311, ...           6.5
...       ...                                                ...           ...
46452   45527                                                 []           3.7
46453  455661  [{'id': 9673, 'name': 'love'}, {'id': 13130, '...           8.3
46454  327237  [{'id': 5565, 'name': 'biography'}, {'id': 102...           6.8
46455   84710           [{'id': 208422, 'name': 'paula dupree'}]           5.8
46456   39562  [{'id': 2064, 'name': 'overweight'}, {'id': 10...           5.0
46457   14008        [{'id': 171803, 'name': 'military school'}]           5.2
46458   44330            [{'id': 169590, 'name': 'police raid'}]           5.0
46459   49279  [{'id': 10124, 'name': 'laboratory'}, {'id': 1...           7.6
46460   44333  [{'id': 2343, 'name': 'magic'}, {'id': 154802,...           6.1
46461   49277  [{'id': 154802, 'name': 'silent film'}, {'id':...           4.5
46462   49271  [{'id': 11320, 'name': 'tenant'}, {'id': 15480...           6.7
46463   44324                                                 []           6.0
46464  122036                                                 []           5.0
46465   14885  [{'id': 3335, 'name': 'halloween'}, {'id': 193...           5.4
46466   49280                                                 []           6.5
46467  106807  [{'id': 5970, 'name': 'wrestling'}, {'id': 607...           6.5
46468  276895                                                 []           0.0
46469  404604  [{'id': 155794, 'name': 'physical abuse'}, {'i...           6.6
46470  420346                                                

In [9]:
#[]削除
keywords_vote=keywords_vote.drop(keywords_vote[keywords_vote.keywords=="[]"].index.values,axis=0)
keywords_vote

id                                           keywords  vote_average
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7
5         949  [{'id': 642, 'name': 'robbery'}, {'id': 703, '...           7.7
6       11860  [{'id': 90, 'name': 'paris'}, {'id': 380, 'nam...           6.2
8        9091  [{'id': 949, 'name': 'terrorist'}, {'id': 1562...           5.5
9         710  [{'id': 701, 'name': 'cuba'}, {'id': 769, 'nam...           6.6
10       9087  [{'id': 833, 'name': 'white house'}, {'id': 84...           6.5
11      12110  [{'id': 3633, 'name': 'dracula'}, {'id': 11931...           5.7
12      21032  [{'id': 1994, 'name': 'wolf'}, {'id': 6411, 'n...           7.1
13      10858  [{'id': 840, 'name': 'usa president'}, {'id': ...           7.1
14       1408  [{'id': 911, 'name': 'exotic island'}, {'id': ...           5.7
15        524  [{'id': 383, 'name': 'poker'}, {'id': 726, 'na...           7.8
16       4584  [{'id': 420, 'name': 'bowling'}, {'id': 818, '...           7.2
17          5  [{'id': 612, 'name': 'hotel'}, {'id': 613, 'na...           6.5
18       9273  [{'id': 409, 'name': 'africa'}, {'id': 1551, '...           6.1
19      11517  [{'id': 380, 'name': 'brother brother relation...           5.4
20       8012  [{'id': 395, 'name': 'gambling'}, {'id': 416, ...           6.4
21       1710  [{'id': 796, 'name': 'police brutality'}, {'id...           6.5
22       9691  [{'id': 271, 'name': 'competition'}, {'id': 44...           6.0
23      12665  [{'id': 7345, 'name': 'dead animal'}, {'id': 1...           6.3
24        451  [{'id': 30, 'name': 'individual'}, {'id': 549,...           7.1
25      16420  [{'id': 497, 'name': 'shakespeare'}, {'id': 93...           7.0
26       9263  [{'id': 1228, 'name': '1970s'}, {'id': 5248, '...           6.6
27      17015  [{'id': 1316, 'name': 'captain'}, {'id': 33384...           7.4
28        902  [{'id': 402, 'name': 'clone'}, {'id': 1566, 'n...           7.6
29      37557  [{'id': 964, 'name': 'servant'}, {'id': 2311, ...           6.5
30       9909  [{'id': 897, 'name': 'rap music'}, {'id': 3436...           6.4
...       ...                                                ...           ...
46419   64197         [{'id': 187056, 'name': 'woman director'}]           6.0
46424  100152         [{'id': 187056, 'name': 'woman director'}]           5.0
46430   75608  [{'id': 2167, 'name': 'ethologist'}, {'id': 19...           5.3
46431  275331  [{'id': 409, 'name': 'africa'}, {'id': 18034, ...           5.0
46433  461297  [{'id': 9663, 'name': 'sequel'}, {'id': 156321...           6.0
46434  258907  [{'id': 14629, 'name': 'experimental film'}, {...           8.0
46435  127762            [{'id': 154802, 'name': 'silent film'}]           5.3
46437   52103        [{'id': 10183, 'name': 'independent film'}]           7.2
46444   45029  [{'id': 720, 'name': 'helicopter'}, {'id': 146...           5.0
46445   36680  [{'id': 310, 'name': 'artificial intelligence'...           5.0
46446  222872  [{'id': 1612, 'name': 'spacecraft'}, {'id': 99...           2.0
46449    5589  [{'id': 3205, 'name': 'fairy tale'}, {'id': 13...           6.1
46453  455661  [{'id': 9673, 'name': 'love'}, {'id': 13130, '...           8.3
46454  327237  [{'id': 5565, 'name': 'biography'}, {'id': 102...           6.8
46455   84710           [{'id': 208422, 'name': 'paula dupree'}]           5.8
46456   39562  [{'id': 2064, 'name': 'overweight'}, {'id': 10...           5.0
46457   14008        [{'id': 171803, 'name': 'military school'}]           5.2
46458   44330            [{'id': 169590, 'name': 'police raid'}]           5.0
46459   49279  [{'id': 10124, 'name': 'laboratory'}, {'id': 1

In [13]:
#ratings 0 削除
keywords_vote=keywords_vote.drop(keywords_vote[keywords_vote.vote_average==0].index.values,axis=0)
keywords_vote

id                                           keywords  vote_average
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7
5         949  [{'id': 642, 'name': 'robbery'}, {'id': 703, '...           7.7
6       11860  [{'id': 90, 'name': 'paris'}, {'id': 380, 'nam...           6.2
7       45325                                                 []           5.4
8        9091  [{'id': 949, 'name': 'terrorist'}, {'id': 1562...           5.5
9         710  [{'id': 701, 'name': 'cuba'}, {'id': 769, 'nam...           6.6
10       9087  [{'id': 833, 'name': 'white house'}, {'id': 84...           6.5
11      12110  [{'id': 3633, 'name': 'dracula'}, {'id': 11931...           5.7
12      21032  [{'id': 1994, 'name': 'wolf'}, {'id': 6411, 'n...           7.1
13      10858  [{'id': 840, 'name': 'usa president'}, {'id': ...           7.1
14       1408  [{'id': 911, 'name': 'exotic island'}, {'id': ...           5.7
15        524  [{'id': 383, 'name': 'poker'}, {'id': 726, 'na...           7.8
16       4584  [{'id': 420, 'name': 'bowling'}, {'id': 818, '...           7.2
17          5  [{'id': 612, 'name': 'hotel'}, {'id': 613, 'na...           6.5
18       9273  [{'id': 409, 'name': 'africa'}, {'id': 1551, '...           6.1
19      11517  [{'id': 380, 'name': 'brother brother relation...           5.4
20       8012  [{'id': 395, 'name': 'gambling'}, {'id': 416, ...           6.4
21       1710  [{'id': 796, 'name': 'police brutality'}, {'id...           6.5
22       9691  [{'id': 271, 'name': 'competition'}, {'id': 44...           6.0
23      12665  [{'id': 7345, 'name': 'dead animal'}, {'id': 1...           6.3
24        451  [{'id': 30, 'name': 'individual'}, {'id': 549,...           7.1
25      16420  [{'id': 497, 'name': 'shakespeare'}, {'id': 93...           7.0
26       9263  [{'id': 1228, 'name': '1970s'}, {'id': 5248, '...           6.6
27      17015  [{'id': 1316, 'name': 'captain'}, {'id': 33384...           7.4
28        902  [{'id': 402, 'name': 'clone'}, {'id': 1566, 'n...           7.6
29      37557  [{'id': 964, 'name': 'servant'}, {'id': 2311, ...           6.5
...       ...                                                ...           ...
46446  222872  [{'id': 1612, 'name': 'spacecraft'}, {'id': 99...           2.0
46449    5589  [{'id': 3205, 'name': 'fairy tale'}, {'id': 13...           6.1
46451   63266                                                 []           3.0
46452   45527                                                 []           3.7
46453  455661  [{'id': 9673, 'name': 'love'}, {'id': 13130, '...           8.3
46454  327237  [{'id': 5565, 'name': 'biography'}, {'id': 102...           6.8
46455   84710           [{'id': 208422, 'name': 'paula dupree'}]           5.8
46456   39562  [{'id': 2064, 'name': 'overweight'}, {'id': 10...           5.0
46457   14008        [{'id': 171803, 'name': 'military school'}]           5.2
46458   44330            [{'id': 169590, 'name': 'police raid'}]           5.0
46459   49279  [{'id': 10124, 'name': 'laboratory'}, {'id': 1...           7.6
46460   44333  [{'id': 2343, 'name': 'magic'}, {'id': 154802,...           6.1
46461   49277  [{'id': 154802, 'name': 'silent film'}, {'id':...           4.5
46462   49271  [{'id': 11320, 'name': 'tenant'}, {'id': 15480...           6.7
46463   44324                                                 []           6.0
46464  122036                                                 []           5.0
46465   14885  [{'id': 3335, 'name': 'halloween'}, {'id': 193...           5.4
46466   49280                                                 []           6.5
46467  106807  [{'id': 5970, 'name': 'wrestling'}, {'id': 607

### keywords_listづくり

In [11]:
#その名の通りstr->list

def make_list_from_str(string):
    if isinstance(string, str):
        result_list = eval(string)
    else:
        result_list = np.nan
    return result_list

In [12]:
#keywords_vote["keywords_list"]がlist化されたkeywords
keywords_vote["keywords_list"] = keywords_vote.keywords.map(make_list_from_str)
keywords_vote = keywords_vote.dropna()
keywords_vote.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30304 entries, 0 to 46478
Data columns (total 4 columns):
id               30304 non-null int64
keywords         30304 non-null object
vote_average     30304 non-null float64
keywords_list    30304 non-null object
dtypes: float64(1), int64(1), object(2)
memory usage: 1.2+ MB


### ダミー変数作成

In [13]:
##ジャンルのdictから、それぞれのidだけを抜き出したリストの作成
def extract_ids(genres_list):
    id_list = []
    for genre in genres_list:
        id_list.append(genre["id"])
    return id_list

In [14]:
##ジャンルのdictから、それぞれのnameだけを抜き出したリストの作成
def extract_names(genres_list):
    id_list = []
    for genre in genres_list:
        id_list.append(genre["name"])
    return id_list

In [15]:
keywords_vote["keywords_name"] = keywords_vote["keywords_list"].map(extract_names)
keywords_vote.head()

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


id                                           keywords  vote_average  \
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7   
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9   
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5   
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1   
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7   

                                       keywords_list  \
0  [{'name': 'jealousy', 'id': 931}, {'name': 'to...   
1  [{'name': 'board game', 'id': 10090}, {'name':...   
2  [{'name': 'fishing', 'id': 1495}, {'name': 'be...   
3  [{'name': 'based on novel', 'id': 818}, {'name...   
4  [{'name': 'baby', 'id': 1009}, {'name': 'midli...   

                                       keywords_name  
0  [jealousy, toy, boy, friendship, friends, riva...  
1  [board game, disappearance, based on children'...  
2  [fishing, best friend, duringcreditsstinger, o...  
3  [based on novel, interracial relationship, sin...  
4  [baby, midlife crisis, confidence, aging, daug...

In [37]:
keywords_vote["keywords_id"] = keywords_vote["keywords_list"].map(extract_ids)
keywords_vote.head()

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


id                                           keywords  vote_average  \
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7   
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9   
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5   
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1   
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7   

                                       keywords_list  \
0  [{'name': 'jealousy', 'id': 931}, {'name': 'to...   
1  [{'name': 'board game', 'id': 10090}, {'name':...   
2  [{'name': 'fishing', 'id': 1495}, {'name': 'be...   
3  [{'name': 'based on novel', 'id': 818}, {'name...   
4  [{'name': 'baby', 'id': 1009}, {'name': 'midli...   

                                       keywords_name  keywords_count  \
0  [jealousy, toy, boy, friendship, friends, riva...               9   
1  [board game, disappearance, based on children'...               6   
2  [fishing, best friend, duringcreditsstinger, o...               4   
3  [based on novel, interracial relationship, sin...               5   
4  [baby, midlife crisis, confidence, aging, daug...               9   

                                         keywords_id  
0  [931, 4290, 5202, 6054, 9713, 9823, 165503, 17...  
1       [10090, 10941, 15101, 33467, 158086, 158091]  
2                      [1495, 12392, 179431, 208510]  
3                  [818, 10131, 14768, 15160, 33455]  
4  [1009, 1599, 2246, 4995, 5600, 10707, 13149, 3...

In [16]:
keywords_vote["keywords_count"] = keywords_vote["keywords_list"].map(lambda x: len(x))
keywords_vote.head()

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


id                                           keywords  vote_average  \
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7   
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9   
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5   
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1   
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7   

                                       keywords_list  \
0  [{'name': 'jealousy', 'id': 931}, {'name': 'to...   
1  [{'name': 'board game', 'id': 10090}, {'name':...   
2  [{'name': 'fishing', 'id': 1495}, {'name': 'be...   
3  [{'name': 'based on novel', 'id': 818}, {'name...   
4  [{'name': 'baby', 'id': 1009}, {'name': 'midli...   

                                       keywords_name  keywords_count  
0  [jealousy, toy, boy, friendship, friends, riva...               9  
1  [board game, disappearance, based on children'...               6  
2  [fishing, best friend, duringcreditsstinger, o...               4  
3  [based on novel, interracial relationship, sin...               5  
4  [baby, midlife crisis, confidence, aging, daug...               9

In [18]:
#空のデータフレーム
keywords_only=pd.DataFrame(index=['index'],columns=['name','num'])
#追加する行をSeriesでつくり、追加していく
for i in keywords_vote.index:
    for j in range(len(keywords_vote.keywords_list[i])):
        if keywords_vote.keywords_list[i][j]["id"] in keywords_only.index:
            keywords_only.num[keywords_vote.keywords_list[i][j]["id"]]+=1
        else:
            df_buffer=pd.DataFrame({'name':keywords_vote.keywords_list[i][j]["name"],'num':1.},index=[keywords_vote.keywords_list[i][j]["id"]])
            keywords_only = keywords_only.append(df_buffer)
            #print(keywords_only.count[0])
            
keywords_only

/Users/minori/.pyenv/versions/anaconda3-4.2.0/lib/python3.5/site-packages/ipykernel/__main__.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


name    num
index                        NaN    NaN
931                     jealousy  194.0
4290                         toy   19.0
5202                         boy   66.0
6054                  friendship  410.0
9713                     friends  167.0
9823                     rivalry   87.0
165503             boy next door    3.0
170722                   new toy    2.0
187065         toy comes to life   17.0
10090                 board game    8.0
10941              disappearance   66.0
15101   based on children's book   31.0
33467                   new home   10.0
158086                   recluse    8.0
158091              giant insect   15.0
1495                     fishing   23.0
12392                best friend  134.0
179431      duringcreditsstinger  439.0
208510                   old men    2.0
818               based on novel  804.0
10131   interracial relationship   30.0
14768              single mother   61.0
15160                    divorce  133.0
33455                chick flick    2.0
1009                        baby  107.0
1599              midlife crisis   38.0
2246                  confidence   18.0
4995                       aging   29.0
5600                    daughter  180.0
...                          ...    ...
41521        first grade teacher    1.0
192168                   dubbing    1.0
199248       death of a musician    1.0
235298             talking snake    1.0
226414                  erotique    1.0
206225          education system    1.0
237085                      vore    1.0
207147         státní bezpečnost    1.0
207148                       stb    1.0
207149            state security    1.0
236966               inescapable    1.0
236967                   remains    1.0
236968            ancient tablet    1.0
227816           provincial life    1.0
236226             anna karenina    1.0
236227            social scandal    1.0
166760       female psychologist    1.0
207902             brain implant    1.0
210381                     nudie    1.0
227888               robespierre    1.0
10487                  adam west    1.0
10488                  burt ward    1.0
166824             gas explosion    1.0
166853                  arm cast    1.0
232748                   campers    1.0
234625         camping equipment    1.0
233372                       emt    1.0
218070          radio controlled    1.0
220150     fictional documentary    1.0
209358                      acab    1.0

[19767 rows x 2 columns]

In [30]:
keywords_only=keywords_only.dropna()

In [31]:
keywords_only.to_csv("keywords_only.csv")

In [36]:
#使われている回数が5回以上のものに絞る
keywords_above=keywords_only[keywords_only.num>5]
keywords_above

name    num
931                         jealousy  194.0
4290                             toy   19.0
5202                             boy   66.0
6054                      friendship  410.0
9713                         friends  167.0
9823                         rivalry   87.0
187065             toy comes to life   17.0
10090                     board game    8.0
10941                  disappearance   66.0
15101       based on children's book   31.0
33467                       new home   10.0
158086                       recluse    8.0
158091                  giant insect   15.0
1495                         fishing   23.0
12392                    best friend  134.0
179431          duringcreditsstinger  439.0
818                   based on novel  804.0
10131       interracial relationship   30.0
14768                  single mother   61.0
15160                        divorce  133.0
1009                            baby  107.0
1599                  midlife crisis   38.0
2246                      confidence   18.0
4995                           aging   29.0
5600                        daughter  180.0
10707   mother daughter relationship  133.0
13149                      pregnancy  140.0
170521                  gynecologist    7.0
642                          robbery  220.0
703                        detective  229.0
...                              ...    ...
15077                   horror movie    7.0
193471                animated short   11.0
210710                     absurdist    6.0
192066                         roast   11.0
207457               celebrity roast   11.0
190483                   bowery boys   16.0
184948                disney channel   14.0
191225                    homophobic    7.0
187864                         human    6.0
189984                         tamil    7.0
193685                     malayalam   29.0
154903                     homophobe    6.0
161151                       fiction    7.0
159551                     pink film    6.0
204298               loss of a child    6.0
155655                 tom and jerry    8.0
197314  sexually transmitted disease    6.0
189152                   dr. kildare    6.0
190901                   perry mason   15.0
157150                         pluto    8.0
203322                       dramedy    6.0
207419                      infected    8.0
196541                       pohádka    7.0
215762                        naruto    7.0
201061                  magical girl    6.0
211467                    trick film   11.0
187305                        peplum   10.0
207573                  cine quinqui    6.0
199356                   veggietales    7.0
231576                  comic meteor    6.0

[4566 rows x 2 columns]

In [64]:
keywords_dummies=pd.DataFrame(columns=[keywords_above.index],index=keywords_vote.index)
keywords_dummies

931    4290   5202   6054   9713   9823   187065 10090  10941  15101   \
0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
6        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
8        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
9        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
10       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
14       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
15       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
16       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
17       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
19       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
20       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
21       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
22       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
23       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
24       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
25       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
26       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
27       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
28       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
29       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
30       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
46419    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46424    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46430    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46431    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46433    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46434    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46435    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46437    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46444    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46445    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46446    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46449    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46453    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46454    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46455    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46456    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46457    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46458    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
46459    NaN    NaN    NaN    NaN    NaN    NaN    NaN  

In [66]:
keywords_dummies=keywords_dummies.fillna(0)

In [68]:
keywords_dummies

931     4290    5202    6054    9713    9823    187065  10090   10941   \
0           0       0       0       0       0       0       0       0       0   
1           0       0       0       0       0       0       0       0       0   
2           0       0       0       0       0       0       0       0       0   
3           0       0       0       0       0       0       0       0       0   
4           0       0       0       0       0       0       0       0       0   
5           0       0       0       0       0       0       0       0       0   
6           0       0       0       0       0       0       0       0       0   
8           0       0       0       0       0       0       0       0       0   
9           0       0       0       0       0       0       0       0       0   
10          0       0       0       0       0       0       0       0       0   
11          0       0       0       0       0       0       0       0       0   
12          0       0       0       0       0       0       0       0       0   
13          0       0       0       0       0       0       0       0       0   
14          0       0       0       0       0       0       0       0       0   
15          0       0       0       0       0       0       0       0       0   
16          0       0       0       0       0       0       0       0       0   
17          0       0       0       0       0       0       0       0       0   
18          0       0       0       0       0       0       0       0       0   
19          0       0       0       0       0       0       0       0       0   
20          0       0       0       0       0       0       0       0       0   
21          0       0       0       0       0       0       0       0       0   
22          0       0       0       0       0       0       0       0       0   
23          0       0       0       0       0       0       0       0       0   
24          0       0       0       0       0       0       0       0       0   
25          0       0       0       0       0       0       0       0       0   
26          0       0       0       0       0       0       0       0       0   
27          0       0       0       0       0       0       0       0       0   
28          0       0       0       0       0       0       0       0       0   
29          0       0       0       0       0       0       0       0       0   
30          0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
46419       0       0       0       0       0       0       0       0       0   
46424       0       0       0       0       0       0       0       0       0   
46430       0       0       0       0       0       0       0       0       0   
46431       0       0       0       0       0       0       0       0       0   
46433       0       0       0       0       0       0       0       0       0   
46434       0       0       0       0       0       0       0       0       0   
46435       0       0       0       0       0       0       0       0       0   
46437       0       0       0       0       0       0       0       0       0   
46444       0       0       0       0       0       0       0       0       0   
46445       0       0       0       0       0       0       0       0       0   
46446       0       0       0       0       0       0       0       0       0   
46449       0       0       0       0       0       0       0       0       0   
46453       0       0       0       0       0       0       0       0       0   
46454       0       0       0       0       0       0       0       0       0   
46455       0       0       0       0       0       0       0       0       0   
46456       0       0       0       0       0       0       0       0       0   
46457       0       0       0       0       0       0       0       0       0   
46458       0       0       0       0 

In [60]:
keywords_vote.keywords_list[0][1]
#in keywords_dummies.columns

{'id': 4290, 'name': 'toy'}

In [71]:
keywords_dummies[keywords_vote.keywords_list[0][1]['id']][8]

0

In [72]:
for i in keywords_vote.index:
    for j in range(len(keywords_vote.keywords_list[i])):
        if keywords_vote.keywords_list[i][j]['id'] in keywords_dummies.columns:
            keywords_dummies[keywords_vote.keywords_list[i][j]['id']][i]=1

In [73]:
keywords_dummies

931     4290    5202    6054    9713    9823    187065  10090   10941   \
0           1       1       1       1       1       1       1       0       0   
1           0       0       0       0       0       0       0       1       1   
2           0       0       0       0       0       0       0       0       0   
3           0       0       0       0       0       0       0       0       0   
4           0       0       0       0       0       0       0       0       0   
5           0       0       0       0       0       0       0       0       0   
6           0       0       0       0       0       0       0       0       0   
8           0       0       0       0       0       0       0       0       0   
9           0       0       0       0       0       0       0       0       0   
10          0       0       0       0       0       0       0       0       0   
11          0       0       0       0       0       0       0       0       0   
12          0       0       0       0       0       0       0       0       0   
13          0       0       0       0       0       0       0       0       0   
14          0       0       0       0       0       0       0       0       0   
15          0       0       0       0       0       0       0       0       0   
16          0       0       0       0       0       0       0       0       0   
17          0       0       0       0       0       0       0       0       0   
18          0       0       0       0       0       0       0       0       0   
19          0       0       0       0       0       0       0       0       0   
20          0       0       0       0       0       0       0       0       0   
21          0       0       0       0       0       0       0       0       0   
22          0       0       0       0       0       1       0       0       0   
23          0       0       0       0       0       0       0       0       0   
24          0       0       0       0       0       0       0       0       0   
25          1       0       0       0       1       0       0       0       0   
26          0       0       0       0       0       0       0       0       0   
27          0       0       0       0       0       0       0       0       0   
28          0       0       0       1       0       0       0       0       0   
29          0       0       0       0       0       0       0       0       0   
30          0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
46419       0       0       0       0       0       0       0       0       0   
46424       0       0       0       0       0       0       0       0       0   
46430       0       0       0       0       0       0       0       0       0   
46431       0       0       0       0       0       0       0       0       0   
46433       0       0       0       0       0       0       0       0       0   
46434       0       0       0       0       0       0       0       0       0   
46435       0       0       0       0       0       0       0       0       0   
46437       0       0       0       0       0       0       0       0       0   
46444       0       0       0       0       0       0       0       0       0   
46445       0       0       0       0       0       0       0       0       0   
46446       0       0       0       0       0       0       0       0       0   
46449       0       0       0       0       0       0       0       0       0   
46453       0       0       0       0       0       0       0       0       0   
46454       0       0       0       0       0       0       0       0       0   
46455       0       0       0       0       0       0       0       0       0   
46456       0       0       0       0       0       0       0       0       0   
46457       0       0       0       0       0       0       0       0       0   
46458       0       0       0       0 

In [76]:
keywords_dummies.to_csv("keywords_dummies.csv")

In [108]:
id_ratings=keywords_vote.drop(["keywords","keywords_list","keywords_count","keywords_name","keywords_id"],axis=1)
id_ratings

id  vote_average
0         862           7.7
1        8844           6.9
2       15602           6.5
3       31357           6.1
4       11862           5.7
5         949           7.7
6       11860           6.2
8        9091           5.5
9         710           6.6
10       9087           6.5
11      12110           5.7
12      21032           7.1
13      10858           7.1
14       1408           5.7
15        524           7.8
16       4584           7.2
17          5           6.5
18       9273           6.1
19      11517           5.4
20       8012           6.4
21       1710           6.5
22       9691           6.0
23      12665           6.3
24        451           7.1
25      16420           7.0
26       9263           6.6
27      17015           7.4
28        902           7.6
29      37557           6.5
30       9909           6.4
...       ...           ...
46419   64197           6.0
46424  100152           5.0
46430   75608           5.3
46431  275331           5.0
46433  461297           6.0
46434  258907           8.0
46435  127762           5.3
46437   52103           7.2
46444   45029           5.0
46445   36680           5.0
46446  222872           2.0
46449    5589           6.1
46453  455661           8.3
46454  327237           6.8
46455   84710           5.8
46456   39562           5.0
46457   14008           5.2
46458   44330           5.0
46459   49279           7.6
46460   44333           6.1
46461   49277           4.5
46462   49271           6.7
46465   14885           5.4
46467  106807           6.5
46469  404604           6.6
46472   84419           6.3
46473  390959           7.0
46474  289923           7.0
46477  439050           4.0
46478  111109           9.0

[30304 rows x 2 columns]

In [113]:
dummies_vote=pd.concat([id_ratings,keywords_dummies],axis=1)
dummies_vote

id  vote_average  931  4290  5202  6054  9713  9823  187065  10090  \
0         862           7.7    1     1     1     1     1     1       1      0   
1        8844           6.9    0     0     0     0     0     0       0      1   
2       15602           6.5    0     0     0     0     0     0       0      0   
3       31357           6.1    0     0     0     0     0     0       0      0   
4       11862           5.7    0     0     0     0     0     0       0      0   
5         949           7.7    0     0     0     0     0     0       0      0   
6       11860           6.2    0     0     0     0     0     0       0      0   
8        9091           5.5    0     0     0     0     0     0       0      0   
9         710           6.6    0     0     0     0     0     0       0      0   
10       9087           6.5    0     0     0     0     0     0       0      0   
11      12110           5.7    0     0     0     0     0     0       0      0   
12      21032           7.1    0     0     0     0     0     0       0      0   
13      10858           7.1    0     0     0     0     0     0       0      0   
14       1408           5.7    0     0     0     0     0     0       0      0   
15        524           7.8    0     0     0     0     0     0       0      0   
16       4584           7.2    0     0     0     0     0     0       0      0   
17          5           6.5    0     0     0     0     0     0       0      0   
18       9273           6.1    0     0     0     0     0     0       0      0   
19      11517           5.4    0     0     0     0     0     0       0      0   
20       8012           6.4    0     0     0     0     0     0       0      0   
21       1710           6.5    0     0     0     0     0     0       0      0   
22       9691           6.0    0     0     0     0     0     1       0      0   
23      12665           6.3    0     0     0     0     0     0       0      0   
24        451           7.1    0     0     0     0     0     0       0      0   
25      16420           7.0    1     0     0     0     1     0       0      0   
26       9263           6.6    0     0     0     0     0     0       0      0   
27      17015           7.4    0     0     0     0     0     0       0      0   
28        902           7.6    0     0     0     1     0     0       0      0   
29      37557           6.5    0     0     0     0     0     0       0      0   
30       9909           6.4    0     0     0     0     0     0       0      0   
...       ...           ...  ...   ...   ...   ...   ...   ...     ...    ...   
46419   64197           6.0    0     0     0     0     0     0       0      0   
46424  100152           5.0    0     0     0     0     0     0       0      0   
46430   75608           5.3    0     0     0     0     0     0       0      0   
46431  275331           5.0    0     0     0     0     0     0       0      0   
46433  461297           6.0    0     0     0     0     0     0       0      0   
46434  258907           8.0    0     0     0     0     0     0       0      0   
46435  127762           5.3    0     0     0     0     0     0       0      0   
46437   52103           7.2    0     0     0     0     0     0       0      0   
46444   45029           5.0    0     0     0     0     0     0       0      0   
46445   36680           5.0    0     0     0     0     0     0       0      0   
46446  222872           2.0    0     0     0     0     0     0       0      0   
46449    5589           6.1    0     0     0     0     0     0       0      0   
46453  455661           8.3    0     0     0     0     0     0       0      0   
46454  327237           6.8    0     0     0     0     0     0       0      0   
46455   84710           5.8    0     0     0     0     0     0       0      0   
46456   39562           5.0    0     0     0     0     0     0       0      0   
46457   14008           5.2    0     0     0     0     0     0       0      0   
46458   44330           5.0    0     0    

In [114]:
#sklearn インポートのヒント　https://github.com/scikit-learn/scikit-learn/issues/6161
from sklearn.model_selection import train_test_split

# モデル
from sklearn import linear_model

# モデルのインスタンス
l_model = linear_model.LinearRegression()
 
# 説明変数に "price" 以外を利用
X = dummies_vote.drop(["vote_average","id"], axis=1)

# 目的変数
Y = dummies_vote["vote_average"]

# 学習データとテストデータ分ける(test_sizeはテストデータの割合)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3,random_state=0)

# モデルのあてはめ
clf = l_model.fit(X_train,y_train)
print("train:",clf.__class__.__name__ ,clf.score(X_train,y_train))
print("test:",clf.__class__.__name__ , clf.score(X_test,y_test))
 
# 偏回帰係数
print(pd.DataFrame({"Name":X.columns,
                    "Coefficients":clf.coef_}).sort_values(by='Coefficients') )
 
# 切片 
print(clf.intercept_)

train: LinearRegression 0.306423900268
test: LinearRegression -0.121126207975
      Coefficients    Name
1606 -8.298563e+12  238239
585  -8.298563e+12  238241
4319 -6.350577e+10  203502
1797 -5.259448e+00  161545
4564 -3.857197e+00  199356
3595 -2.197628e+00    1472
4381 -2.143349e+00    1477
2162 -2.114155e+00  156237
3449 -1.947073e+00  189400
2379 -1.895874e+00    1828
4535 -1.842919e+00   14620
3156 -1.723810e+00  157238
4277 -1.716504e+00    2834
3431 -1.691001e+00  157677
3074 -1.665299e+00  158072
3940 -1.664191e+00    6310
4216 -1.635323e+00  209474
4334 -1.557993e+00  191146
3772 -1.525298e+00  161351
3568 -1.523395e+00  158199
2809 -1.523341e+00   10328
2700 -1.517217e+00  158241
3125 -1.516642e+00  180824
3730 -1.513334e+00    6954
3330 -1.485674e+00  162328
715  -1.476898e+00   41175
3646 -1.468210e+00    2597
3549 -1.465080e+00   33595
2826 -1.447323e+00    4953
3809 -1.425235e+00  170375
...            ...     ...
2558  1.498832e+00    6399
4507  1.503300e+00  175322
2144

In [115]:
print(pd.DataFrame({"Name":X.columns,
                    "Coefficients":clf.coef_}).sort_values(by='Coefficients').sort_values("Coefficients") )

      Coefficients    Name
1606 -8.298563e+12  238239
585  -8.298563e+12  238241
4319 -6.350577e+10  203502
1797 -5.259448e+00  161545
4564 -3.857197e+00  199356
3595 -2.197628e+00    1472
4381 -2.143349e+00    1477
2162 -2.114155e+00  156237
3449 -1.947073e+00  189400
2379 -1.895874e+00    1828
4535 -1.842919e+00   14620
3156 -1.723810e+00  157238
4277 -1.716504e+00    2834
3431 -1.691001e+00  157677
3074 -1.665299e+00  158072
3940 -1.664191e+00    6310
4216 -1.635323e+00  209474
4334 -1.557993e+00  191146
3772 -1.525298e+00  161351
3568 -1.523395e+00  158199
2809 -1.523341e+00   10328
2700 -1.517217e+00  158241
3125 -1.516642e+00  180824
3730 -1.513334e+00    6954
3330 -1.485674e+00  162328
715  -1.476898e+00   41175
3646 -1.468210e+00    2597
3549 -1.465080e+00   33595
2826 -1.447323e+00    4953
3809 -1.425235e+00  170375
...            ...     ...
2558  1.498832e+00    6399
4507  1.503300e+00  175322
2144  1.508745e+00   33553
2515  1.516370e+00    3068
2937  1.526535e+00  191786
3

id 606 orcs 10回
id 9550 doc holiday 6回
id 10037　le mans 7回
id 11318　landlord 9回
id 3417 wormhole 7回
id 233981 formula one 6回
id 430 venezuela 8回

1797 -5.259448e+00  161545　alcoholic father 6
4564 -3.857197e+00  199356 veggietales 7
3595 -2.197628e+00    1472 starships 6
4381 -2.143349e+00    1477 world cup 6
2162 -2.114155e+00  156237 jewery store


In [3]:
keywords_dummies=pd.read_csv("keywords_dummies.csv")
keywords_dummies.head()

Unnamed: 0  931  4290  5202  6054  9713  9823  187065  10090  10941  \
0           0    1     1     1     1     1     1       1      0      0   
1           1    0     0     0     0     0     0       0      1      1   
2           2    0     0     0     0     0     0       0      0      0   
3           3    0     0     0     0     0     0       0      0      0   
4           4    0     0     0     0     0     0       0      0      0   

    ...    203322  207419  196541  215762  201061  211467  187305  207573  \
0   ...         0       0       0       0       0       0       0       0   
1   ...         0       0       0       0       0       0       0       0   
2   ...         0       0       0       0       0       0       0       0   
3   ...         0       0       0       0       0       0       0       0   
4   ...         0       0       0       0       0       0       0       0   

   199356  231576  
0       0       0  
1       0       0  
2       0       0  
3       0       0  
4       0       0  

[5 rows x 4567 columns]

In [18]:
keywords_only=pd.read_csv("keywords_only.csv",index="Unnamed: 0")
keywords_only.head()

TypeError: parser_f() got an unexpected keyword argument 'index'

In [15]:
keywords_vote.head()

id                                           keywords  vote_average
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...           7.7
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...           6.9
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...           6.5
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...           6.1
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...           5.7

In [ ]:
#sklearn インポートのヒント　https://github.com/scikit-learn/scikit-learn/issues/6161
from sklearn.model_selection import train_test_split

# モデル
from sklearn import linear_model

# モデルのインスタンス
l_model = linear_model.LinearRegression()
 
# 説明変数に "price" 以外を利用
X = dummies_vote.drop(["vote_average","id"], axis=1)

# 目的変数
Y = dummies_vote["vote_average"]

# 学習データとテストデータ分ける(test_sizeはテストデータの割合)
X1_train, X1_test, y1_train, y1_test = train_test_split(X, Y, test_size=0.5,random_state=0)

# モデルのあてはめ
clf1 = l_model.fit(X1_train,y1_train)
print("train:",clf.__class__.__name__ ,clf.score(X1_train,y1_train))
print("test:",clf.__class__.__name__ , clf.score(X1_test,y1_test))
 
# 偏回帰係数
print(pd.DataFrame({"Name":X1.columns,
                    "Coefficients":clf1.coef_}).sort_values(by='Coefficients') )
 
# 切片 
print(clf1.intercept_)

In [ ]:
above5=pd.DataFrame({"Name":X.columns,
                    "Coefficients":clf.coef_}).sort_values(by='Coefficients')
pd.merge(above5,keywords_only,left_on= "Name",right_on="index")

In [ ]:
a

In [123]:
#よく使われてるkeywordのcolumnのダミーが全て０の行の消去
dummies_vote_fixed=dummies_vote.applymap(lambda x : np.nan if x==0 else x).dropna(axis=0,how='all')
dummies_vote_fixed

id  vote_average  931  4290  5202  6054  9713  9823  187065  10090  \
0         862           7.7  1.0   1.0   1.0   1.0   1.0   1.0     1.0    NaN   
1        8844           6.9  NaN   NaN   NaN   NaN   NaN   NaN     NaN    1.0   
2       15602           6.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
3       31357           6.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
4       11862           5.7  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
5         949           7.7  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
6       11860           6.2  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
8        9091           5.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
9         710           6.6  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
10       9087           6.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
11      12110           5.7  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
12      21032           7.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
13      10858           7.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
14       1408           5.7  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
15        524           7.8  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
16       4584           7.2  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
17          5           6.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
18       9273           6.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
19      11517           5.4  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
20       8012           6.4  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
21       1710           6.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
22       9691           6.0  NaN   NaN   NaN   NaN   NaN   1.0     NaN    NaN   
23      12665           6.3  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
24        451           7.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
25      16420           7.0  1.0   NaN   NaN   NaN   1.0   NaN     NaN    NaN   
26       9263           6.6  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
27      17015           7.4  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
28        902           7.6  NaN   NaN   NaN   1.0   NaN   NaN     NaN    NaN   
29      37557           6.5  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
30       9909           6.4  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
...       ...           ...  ...   ...   ...   ...   ...   ...     ...    ...   
46419   64197           6.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46424  100152           5.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46430   75608           5.3  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46431  275331           5.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46433  461297           6.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46434  258907           8.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46435  127762           5.3  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46437   52103           7.2  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46444   45029           5.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46445   36680           5.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46446  222872           2.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46449    5589           6.1  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46453  455661           8.3  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46454  327237           6.8  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46455   84710           5.8  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46456   39562           5.0  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46457   14008           5.2  NaN   NaN   NaN   NaN   NaN   NaN     NaN    NaN   
46458   44330           5.0  NaN   NaN   N